### DISCUSSION TITLE
skynet如何进行流量控制？

### DISCUSSION THREAD

**Discussion by chengf2018** [cloudwu/skynet#1917] [Q&A]
看了skynet提供了pause/resume机制，但是如何精确的控制socket的流量呢？
例如：将socket控制在每秒不超过1024kb。

**Comment by chengf2018** [comment]
流量控制指的是read的流量，不是write的

**Comment by 253980289** [comment]
可以考虑在应用层read后弄一个滑动窗口统计收到的字节数，当速率大于你设定的速率时，就调用socket.pause，可以参考skynet的socket.lua自身源码中对pause和resume的实现，只不过其仅按缓冲区大小来控制的。

**Comment by fanyh** [comment]
[goscon](https://github.com/ejoy/goscon)

**Comment by jixiang4223** [comment]
令牌桶
